### PR TITLE
Clean up the feed in HomeActivity - Step 3

### PR DESCRIPTION
**Pull request by florina-muntenescu**
Move logic from HomeActivity to ViewModel:
* Handling of the removal/deactivation of a source
* Requesting data loading at every activity onCreate - moved to init of the ViewModel
* Move check for blank query to HomeViewModel

Update the weighting algorithm to allow multiple types in the list.

### CODE REVIEW

File: app/src/main/java/io/plaidapp/ui/HomeViewModel.kt [hunk 1]
```diff
@@ -88,10 +99,15 @@ class HomeViewModel(
         }
     }
 
+    // TODO - find a better solution
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeViewModel.kt]: Issue number

File: app/src/main/java/io/plaidapp/ui/HomeViewModel.kt [hunk 2]
```diff
@@ -88,10 +99,15 @@ class HomeViewModel(
...
+    var columns = 2
+
     init {
         sourcesRepository.registerFilterChangedCallback(filtersChangedCallbacks)
+        dataManager.setOnDataLoadedCallback(onDataLoadedCallback)
```
Comment (nickbutcher) [inline app/src/main/java/io/plaidapp/ui/HomeViewModel.kt]: So we have `onDataLoadedCallbacks` and `dataLoadingCallbacks` objects in this class. Wonder if we could clarify the naming to better differentiate them?

File: app/src/test/java/io/plaidapp/TestData.kt [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+import io.plaidapp.core.dribbble.data.api.model.Images
+import io.plaidapp.core.dribbble.data.api.model.Shot
+import io.plaidapp.core.dribbble.data.api.model.User
+import io.plaidapp.core.producthunt.data.api.model.Post
+import io.plaidapp.core.ui.filter.SourceUiModel
+import java.util.Date
+import java.util.GregorianCalendar
+
+val designerNewsSource = Source.DesignerNewsSearchSource(
+    "query",
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/TestData.kt]: Named params for readability.

Comment (nickbutcher) [reply app/src/test/java/io/plaidapp/TestData.kt]: Throughout for test data.

File: app/src/test/java/io/plaidapp/TestData.kt [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    avatarUrl = "www.prettyplaid.nb"
+)
+
+val shot = Shot(
+    id = 1L,
+    title = "Foo Nick",
+    description = "",
+    images = Images(),
+    user = player
+).also {
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/TestData.kt]:
```suggestion
).apply {
```

File: app/src/test/java/io/plaidapp/TestData.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    downvotes = emptyList()
+)
+
+val story = Story(
+    id = storyId,
+    title = "Plaid 2.0 was released",
+    createdAt = createdDate,
+    userId = userId,
+    links = storyLinks
+).also {
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/TestData.kt]:
```suggestion
).apply {
```

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt [hunk 6]
```diff
@@ -126,7 +134,11 @@ class HomeViewModelTest {
         val homeViewModel = createViewModel()
 
         // When adding a Designer News source
-        homeViewModel.addSources(designerNewsSource.query, isDribbble = false, isDesignerNews = true)
+        homeViewModel.addSources(
+            designerNewsSource.query,
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]:
```suggestion
            query = designerNewsSource.query,
```

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt [hunk 7]
```diff
@@ -259,41 +271,42 @@ class HomeViewModelTest {
...
-        val homeViewModel = createViewModel()
+        // Given a view model with feed data
+        val homeViewModel = createViewModelWithFeedData(listOf(post, shot, story))
         verify(sourcesRepository).registerFilterChangedCallback(
             capture(filtersChangedCallback)
         )
+        val initialFeed = LiveDataTestUtil.getValue(homeViewModel.feed)
 
         // When an active source was changed
         val activeSource = Source.DribbbleSearchSource("dribbble", true)
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: Can we copy `dribbbleSource` test data to make the pertinent data obvious? i.e.

`val activeSource = dribbbleSource.copy(active = true)`

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt [hunk 8]
```diff
@@ -302,9 +315,9 @@ class HomeViewModelTest {
         val inactiveSource = Source.DribbbleSearchSource("dribbble", false)
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: Ditto above.

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt [hunk 9]
```diff
@@ -335,6 +348,39 @@ class HomeViewModelTest {
...
+    private fun createViewModelWithFeedData(feedData: List<PlaidItem>): HomeViewModel {
+        val homeViewModel = createViewModel()
+        verify(dataManager).setOnDataLoadedCallback(capture(dataLoadedCallback))
+
+        // When data loaded return feedData
+        dataLoadedCallback.value.onDataLoaded(feedData)
+
+        return homeViewModel
+    }
+
```
Comment (nickbutcher) [inline app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: We're not testing that popular items are expanded which happens when the feed changes. File an issue to come back to this?

File: core/src/main/java/io/plaidapp/core/data/DataManager.java [hunk 10]
```diff
@@ -72,13 +71,15 @@ public DataManager(OnDataLoadedCallback<List<? extends PlaidItem>> onDataLoadedC
...
 
-    private void setOnDataLoadedCallback(
+    public void setOnDataLoadedCallback(
             OnDataLoadedCallback<List<? extends PlaidItem>> onDataLoadedCallback) {
         this.onDataLoadedCallback = onDataLoadedCallback;
     }
 
     private void onDataLoaded(List<? extends PlaidItem> data) {
-        onDataLoadedCallback.onDataLoaded(data);
+        if(onDataLoadedCallback != null) {
```
Comment (nickbutcher) [inline core/src/main/java/io/plaidapp/core/data/DataManager.java]:
```suggestion
        if (onDataLoadedCallback != null) {
```

File: app/src/main/java/io/plaidapp/ui/HomeViewModel.kt (hunk 2 above)
Comment (florina-muntenescu) [reply app/src/main/java/io/plaidapp/ui/HomeViewModel.kt]: I'm planning to remove at least one of them in a next PR. 
I don't think we should use callbacks at all. 
Is it worth changing it now?

File: app/src/test/java/io/plaidapp/TestData.kt (hunk 3 above)
Comment (florina-muntenescu) [reply app/src/test/java/io/plaidapp/TestData.kt]: added to the data classes.
Some others, like the Source here, is still java

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt (hunk 8 above)
Comment (florina-muntenescu) [reply app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: java :)

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt (hunk 9 above)
Comment (florina-muntenescu) [reply app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: #653 created

File: app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt (hunk 7 above)
Comment (florina-muntenescu) [reply app/src/test/java/io/plaidapp/ui/HomeViewModelTest.kt]: would love to. but sources are still java

**Comment by nickbutcher** [comment]
Thanks for the changes 👍

**Pull request merged**